logdevice/ops/ldquery/tables/Directories.cpp

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+#include "logdevice/ops/ldquery/tables/Directories.h"
+
+#include <queue>
+#include <folly/Conv.h>
+#include <folly/json.h>
+
+#include "../Table.h"
+#include "../Utils.h"
+#include "logdevice/common/configuration/Configuration.h"
+#include "logdevice/common/configuration/ReplicationProperty.h"
+#include "logdevice/common/configuration/UpdateableConfig.h"
+#include "logdevice/common/debug.h"
+#include "logdevice/lib/ClientImpl.h"
+
+
+namespace facebook {
+  namespace logdevice {
+    namespace ldquery {
+      namespace tables {
+
+TableColumns Directories::getColumns() const {
+  return {
+      {"path", DataType::TEXT, "Name of the parent directory."},
+      {"name", DataType::TEXT, "Name of the directory."},
+      {"replication_property",
+       DataType::TEXT,
+       "Replication property configured for this directory."},
+      {"synced_copies",
+       DataType::INTEGER,
+       "Number of copies that must be acknowledged by storage nodes are synced "
+       "to disk before the record is acknowledged to the client as fully "
+       "appended."},
+      {"max_writes_in_flight",
+       DataType::INTEGER,
+       "The largest number of records not released for delivery that the "
+       "sequencer allows to be outstanding."},
+      {"backlog_duration_sec",
+       DataType::INTEGER,
+       "Time-based retention of records of logs in that directory.  If null or "
+       "zero, this directory does not use time-based retention."},
+      {"storage_set_size",
+       DataType::INTEGER,
+       "Size of the storage set for logs in that directory.  The storage set "
+       "is the set of shards that may hold data for a log."},
+      {"delivery_latency",
+       DataType::INTEGER,
+       "For logs in that directory, maximum amount of time that we can delay "
+       "delivery of newly written records.  This option increases delivery "
+       "latency but improves server and client performance."},
+      {"scd_enabled",
+       DataType::INTEGER,
+       "Indicates whether the Single Copy Delivery optimization is enabled for "
+       "this directory.  This efficiency optimization allows only one copy of "
+       "each record to be served to readers."},
+      {"custom_fields",
+       DataType::TEXT,
+       "Custom text field provided by the user."},
+  };
+}
+
+std::shared_ptr<TableData> Directories::getData(QueryContext& ctx) {
+  auto result = std::make_shared<TableData>();
+  auto full_client = ld_ctx_->getFullClient();
+
+  ld_check(full_client);
+  ClientImpl* client_impl = static_cast<ClientImpl*>(full_client.get());
```

Comment:
I don't think you need the downcast here. All the APIs seems exposed in the main client API.

Reply:
will remove the downcast